putput/pipeline.py

```
@@ -517,75 +511,63 @@ def _expand(self,
-                expanded_utterance_patterns, _ = expand_utterance_patterns_ranges_and_groups((utterance_pattern,),
-                                                                                             groups_map)
-                for expanded_utterance_pattern in expanded_utterance_patterns:
-                    expanded_map[', '.join(expanded_utterance_pattern)] = hooks
-        return expanded_map
+@overload
+def _execute_hooks(tokens: Sequence[str],
+                   args: Tuple[Sequence[Sequence[str]], Sequence[str], Sequence[Tuple[str, int]]],
+                   hooks_map: _E_H_MAP,
+                   ) -> Tuple[Sequence[Sequence[str]], Sequence[str], Sequence[Tuple[str, int]]]:
+    # pylint: disable=unused-argument
+    pass # pragma: no cover
+
+@overload
+def _execute_hooks(tokens: Sequence[str],
+                   args: Tuple[str, Sequence[str], Sequence[str]],
+                   hooks_map: _C_H_MAP,
+                   ) -> Any:
+    # pylint: disable=unused-argument
+    pass # pragma: no cover
+
+def _execute_hooks(tokens: Sequence[str],
+                   args: Union[Tuple[Sequence[Sequence[str]], Sequence[str], Sequence[Tuple[str, int]]],
+                               Tuple[str, Sequence[str], Sequence[str]]],
+                   hooks_map: Union[_E_H_MAP, _C_H_MAP]
+                   ) -> Union[Tuple[Sequence[Sequence[str]], Sequence[str], Sequence[Tuple[str, int]]], Any]:
+    key = ', '.join(tokens)
+    if key not in hooks_map:
+        key = 'DEFAULT'
+    if key in hooks_map:
+        args = reduce(lambda args, hook: hook(*args), hooks_map[key], args)
+    return args
+
+def _get_combo_options(tokens: Sequence[str],
+                       combo_options_map: Mapping[str, ComboOptions]
+                       ) -> Optional[ComboOptions]:
+    # pylint: disable=no-self-use
+    options_map = {} # type: Dict[str, ComboOptions]
+    options_map.update(combo_options_map)
+    key = ', '.join(tokens)
```

should we throw an error here to let the user know that the key they specified is not valid?

---

instead of just going with DEFAULT

---

On the one hand, I think it could be frustrating as we prototype with the patterns to keep getting errors everytime one of the keys is not used (imagine changing the pattern defintion file and code, going back and forth). On the other hand, this suggestion would help avoid bugs. As this PR doesn't really address that problem, I will merge it and then we can discuss it